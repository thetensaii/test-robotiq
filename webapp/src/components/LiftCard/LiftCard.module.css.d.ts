declare const styles: {
  readonly "liftCard": string;
  readonly "form": string;
  readonly "formField": string;
  readonly "submitButton": string;
};
export = styles;

