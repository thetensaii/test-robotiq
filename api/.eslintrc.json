{
	"env": {
		"es2021": true,
		"node": true
	},
	"extends": ["prettier", "eslint:recommended", "plugin:@typescript-eslint/recommended"],
	"parser": "@typescript-eslint/parser",
	"parserOptions": {
		"ecmaVersion": "latest",
		"sourceType": "module"
	},
	"plugins": ["prettier", "@typescript-eslint"],
	"rules": {
		"prettier/prettier": ["error"],
		"indent": ["error", "tab"],
		"linebreak-style": ["error", "unix"],
		"quotes": ["error", "single", { "avoidEscape": true }],
		"semi": ["error", "always"],
		"no-extra-semi": ["error"],
		"prefer-const": ["warn"]
	}
}
