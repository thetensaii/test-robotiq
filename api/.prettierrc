{
	"bracketSpacing": true,
	"useTabs": true,
	"tabWidth": 4,
	"singleQuote": true,
	"quoteProps": "consistent",
	"printWidth": 100
}
