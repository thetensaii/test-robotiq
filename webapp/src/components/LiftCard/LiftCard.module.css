.liftCard {
	display: flex;
	flex-direction: column;
	align-items: center;

	background-color: var(--primary-color);
	color: aliceblue;
	font-weight: 600;

	margin: 16px;
	padding: 16px 32px;
	border-radius: 8px;

	box-shadow: 1px 6px 4px rgba(0, 0, 0, 0.28);
}

.form {
	display: flex;
	flex-direction: column;
	justify-content: center;
	width: 70%;
}

.formField {
	display: flex;
	justify-content: space-between;
	align-items: center;
	margin: 5px;
}

.submitButton {
	background-color: white;
	color: var(--primary-color);

	font-weight: 600;
	font-size: medium;
	padding: 5px 10px;

	border-style: none;
	border-radius: 10px;
}

.submitButton:hover {
	cursor: pointer;
	background-color: #ddd;
}
