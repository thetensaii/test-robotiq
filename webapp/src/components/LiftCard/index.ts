export { LiftCard as default } from './LiftCard';
