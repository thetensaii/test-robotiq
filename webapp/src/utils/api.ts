import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
import config from '../config';
export const apiGetRequest = async (
	endpoint: string,
	requestConfig: AxiosRequestConfig = {}
): Promise<AxiosResponse> => {
	const response: AxiosResponse = await axios.get(config.API_URL + endpoint, requestConfig);
	return response;
};
export const apiPostRequest = async (
	endpoint: string,
	data: any,
	requestConfig: AxiosRequestConfig = {}
): Promise<AxiosResponse> => {
	const response: AxiosResponse = await axios.post(
		config.API_URL + endpoint,
		data,
		requestConfig
	);

	return response;
};

export const enum ApiPaths {
	LIFTS = '/lifts',
	TRAVELS = '/travels',
}
