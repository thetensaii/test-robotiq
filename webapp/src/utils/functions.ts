export const getAllNumbersBetween = (min: number, max: number): number[] => {
	const numbers = [];

	for (let i = min; i < max + 1; i++) {
		numbers.push(i);
	}

	return numbers;
};

export const getFormData: (form: HTMLFormElement) => { [k: string]: FormDataEntryValue } = (
	form
) => {
	const formData: FormData = new FormData(form);
	return Object.fromEntries(formData);
};
