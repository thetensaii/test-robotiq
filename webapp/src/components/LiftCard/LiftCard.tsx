import React from 'react';
import { getFormData } from '../../utils/functions';
import styles from './LiftCard.module.css';
import Select from '../Select';
import { useLift } from 'hooks/useLift';
import { useAlert } from 'hooks/useAlert';
import AlertMessageBox, { AlertMessageBoxTypes } from 'components/AlertMessageBox';
import axios from 'axios';
import { ImArrowRight2 } from 'react-icons/im';
export type LiftCardProps = {
	id: number;
	name: string;
	currentLocation: number;
	minFloor: number;
	maxFloor: number;
};

export const LiftCard: React.FC<LiftCardProps> = ({
	id,
	name,
	currentLocation,
	minFloor,
	maxFloor,
}) => {
	const [alert, updateAlert, removeAlert] = useAlert();

	const [lift, floors, createTravel] = useLift({
		id: id,
		name: name,
		currentLocation: currentLocation,
		minFloor: minFloor,
		maxFloor: maxFloor,
	});

	const onTravelSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
		e.preventDefault();
		removeAlert();

		const createTravelForm = e.currentTarget;
		const createTravelFormDataRaw = getFormData(createTravelForm);

		const travel = {
			sourceFloor: Number(createTravelFormDataRaw.sourceFloor),
			targetFloor: Number(createTravelFormDataRaw.targetFloor),
		};

		try {
			await createTravel(travel.sourceFloor, travel.targetFloor);
			updateAlert(
				AlertMessageBoxTypes.success,
				`${lift.name} took you from ${travel.sourceFloor} floor to ${travel.targetFloor}`
			);
		} catch (error) {
			if (axios.isAxiosError(error) && error.response) {
				updateAlert(AlertMessageBoxTypes.danger, error.response.data);
			} else if (error instanceof Error) {
				updateAlert(AlertMessageBoxTypes.danger, 'Une erreur a été rencontrée');
			}
		}
	};

	return (
		<div className={styles.liftCard}>
			<h3>{name}</h3>
			<p>Location : {lift.currentLocation}</p>
			<div>
				This lift goes from {lift.minFloor} to {lift.maxFloor}
			</div>
			<form aria-label="form" className={styles.form} onSubmit={onTravelSubmit}>
				<div className={styles.formField}>
					<Select
						name="sourceFloor"
						options={floors.map((f) => ({ value: f, label: f }))}
					/>

					<ImArrowRight2 />
					<Select
						name="targetFloor"
						options={floors.map((f) => ({ value: f, label: f }))}
					/>
				</div>
				<button type="submit" className={styles.submitButton}>
					Travel
				</button>
			</form>
			{alert && (
				<AlertMessageBox type={alert.type} closeAlert={removeAlert}>
					{alert.message}
				</AlertMessageBox>
			)}
		</div>
	);
};
