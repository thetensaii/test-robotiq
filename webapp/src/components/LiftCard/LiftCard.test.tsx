import React from 'react';
import { render, screen, act } from '@testing-library/react';
import { LiftCard } from './LiftCard';
import axios from 'axios';
import userEvent from '@testing-library/user-event';
import config from 'config';
import { ApiPaths } from 'utils/api';

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('LiftCard component tests', () => {
	it('should render LiftCard', () => {
		const { container } = render(
			<LiftCard id={6} name={'Lift test'} currentLocation={0} minFloor={-2} maxFloor={8} />
		);
		expect(container).toMatchSnapshot();
	});

	it('should render lift name', () => {
		render(
			<LiftCard id={6} name={'Lift test'} currentLocation={0} minFloor={-2} maxFloor={8} />
		);

		const nameElement = screen.getByRole('heading', { name: /lift test/i });

		expect(nameElement).toBeInTheDocument();
	});

	it('should render lift location', () => {
		render(
			<LiftCard id={6} name={'Lift test'} currentLocation={0} minFloor={-2} maxFloor={8} />
		);

		const locationElement = screen.getByText(/location : 0/i);
		expect(locationElement).toBeInTheDocument();
	});

	it('should render lift floors', () => {
		render(
			<LiftCard id={6} name={'Lift test'} currentLocation={0} minFloor={-2} maxFloor={8} />
		);

		const floorsElement = screen.getByText(/this lift goes from -2 to 8/i);
		expect(floorsElement).toBeInTheDocument();
	});

	it('should render travel form', () => {
		render(
			<LiftCard id={6} name={'Lift test'} currentLocation={0} minFloor={-2} maxFloor={8} />
		);

		const form: HTMLFormElement = screen.getByRole('form');
		const selectsElement = screen.getAllByRole('combobox');
		const submitElement = screen.getByRole('button', { name: /travel/i });

		expect(form).toBeInTheDocument();

		expect(selectsElement.length).toBe(2);
		expect(selectsElement[0]).toBeInTheDocument();
		expect(selectsElement[1]).toBeInTheDocument();

		expect(submitElement).toBeInTheDocument();
	});

	it('should set lift location', async () => {
		mockedAxios.post.mockResolvedValueOnce({
			data: {},
			status: 201,
			statusText: '',
			headers: {},
			config: {},
		});

		const user = userEvent.setup();
		const alertContainer = document.createElement('div');
		alertContainer.setAttribute('id', 'alert');

		render(
			<LiftCard id={6} name={'Lift test'} currentLocation={0} minFloor={-2} maxFloor={8} />,
			{
				container: document.body.appendChild(alertContainer),
			}
		);

		const submitElement = screen.getByRole('button', { name: /travel/i });
		const locationElement = screen.getByText(/location : 0/i);
		const selectsElement = screen.getAllByRole('combobox');
		const optionsElement: HTMLOptionElement[] = screen.getAllByRole('option');

		// Initial Test
		expect(locationElement).toBeInTheDocument();

		// Select 2nd floor for targetFloor
		await user.selectOptions(selectsElement[1], optionsElement[15]);
		expect(optionsElement[15].selected).toBeTruthy();
		expect(mockedAxios).toHaveBeenCalledTimes(0);

		// Send request
		await act(async () => {
			await user.click(submitElement);
		});

		expect(mockedAxios.post).toHaveBeenCalledTimes(1);
		expect(mockedAxios.post).toHaveBeenCalledWith(
			config.API_URL + ApiPaths.TRAVELS,
			{
				liftID: 6,
				sourceFloor: -2,
				targetFloor: 2,
			},
			{}
		);

		expect(locationElement).toBeInTheDocument();
		expect(locationElement).toHaveTextContent(/location : 2/i);
	});

	it('should not set lift location', async () => {
		mockedAxios.post.mockRejectedValueOnce({
			data: {},
			status: 201,
			statusText: '',
			headers: {},
			config: {},
		});

		const user = userEvent.setup();
		const alertContainer = document.createElement('div');
		alertContainer.setAttribute('id', 'alert');

		render(
			<LiftCard id={6} name={'Lift test'} currentLocation={0} minFloor={-2} maxFloor={8} />,
			{
				container: document.body.appendChild(alertContainer),
			}
		);

		const submitElement = screen.getByRole('button', { name: /travel/i });
		const locationElement = screen.getByText(/location : 0/i);

		// Initial Test
		expect(locationElement).toBeInTheDocument();

		// Send request
		await act(async () => {
			await user.click(submitElement);
		});

		expect(mockedAxios.post).toHaveBeenCalledTimes(1);
		expect(mockedAxios.post).toHaveBeenCalledWith(
			config.API_URL + ApiPaths.TRAVELS,
			{
				liftID: 6,
				sourceFloor: -2,
				targetFloor: -2,
			},
			{}
		);

		expect(locationElement).toBeInTheDocument();
		expect(locationElement).toHaveTextContent(/location : 0/i);
	});
});
