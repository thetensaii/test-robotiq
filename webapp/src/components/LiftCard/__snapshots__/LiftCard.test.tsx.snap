// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiftCard component tests should render LiftCard 1`] = `
<div>
  <div
    class="liftCard"
  >
    <h3>
      Lift test
    </h3>
    <p>
      Location : 
      0
    </p>
    <div>
      This lift goes from 
      -2
       to 
      8
    </div>
    <form
      aria-label="form"
      class="form"
    >
      <div
        class="formField"
      >
        <select
          id="sourceFloor"
          name="sourceFloor"
        >
          <option
            value="-2"
          >
            -2
          </option>
          <option
            value="-1"
          >
            -1
          </option>
          <option
            value="0"
          >
            0
          </option>
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
          <option
            value="3"
          >
            3
          </option>
          <option
            value="4"
          >
            4
          </option>
          <option
            value="5"
          >
            5
          </option>
          <option
            value="6"
          >
            6
          </option>
          <option
            value="7"
          >
            7
          </option>
          <option
            value="8"
          >
            8
          </option>
        </select>
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          version="1.1"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.707 13.707l5-5c0.391-0.39 0.391-1.024 0-1.414l-5-5c-0.391-0.391-1.024-0.391-1.414 0s-0.391 1.024 0 1.414l3.293 3.293h-9.586c-0.552 0-1 0.448-1 1s0.448 1 1 1h9.586l-3.293 3.293c-0.195 0.195-0.293 0.451-0.293 0.707s0.098 0.512 0.293 0.707c0.391 0.391 1.024 0.391 1.414 0z"
          />
        </svg>
        <select
          id="targetFloor"
          name="targetFloor"
        >
          <option
            value="-2"
          >
            -2
          </option>
          <option
            value="-1"
          >
            -1
          </option>
          <option
            value="0"
          >
            0
          </option>
          <option
            value="1"
          >
            1
          </option>
          <option
            value="2"
          >
            2
          </option>
          <option
            value="3"
          >
            3
          </option>
          <option
            value="4"
          >
            4
          </option>
          <option
            value="5"
          >
            5
          </option>
          <option
            value="6"
          >
            6
          </option>
          <option
            value="7"
          >
            7
          </option>
          <option
            value="8"
          >
            8
          </option>
        </select>
      </div>
      <button
        class="submitButton"
        type="submit"
      >
        Travel
      </button>
    </form>
  </div>
</div>
`;
